IdentityCore/src/controllers/ios/MSIDBrokerController.m

```diff
@@ -0,0 +1,293 @@
+
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(appEnteredForeground:)
+                                                 name:UIApplicationWillEnterForegroundNotification
+                                               object:nil];
+#endif
+}
+
+#if !AD_BROKER
+
++ (void)appEnteredForeground:(NSNotification *)notification
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self
+                                                    name:UIApplicationWillEnterForegroundNotification
+                                                  object:nil];
+
+    // Now that we know we've just been woken up from having been in the background we can start listening for
+    // ApplicationDidBecomeActive without having to worry about something else causing it to get hit between
+    // now and openURL:, if we're indeed getting a URL.
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(checkTokenResponse:)
+                                                 name:UIApplicationDidBecomeActiveNotification object:nil];
+}
+
++ (void)checkTokenResponse:(NSNotification *)notification
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self
+                                                    name:UIApplicationDidBecomeActiveNotification
+                                                  object:nil];
+
+    if ([self.class currentBrokerController])
+    {
+        NSError *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorBrokerResponseNotReceived, @"application did not receive response from broker.", nil, nil, nil, nil, nil);
+
+        MSIDBrokerController *brokerController = [self.class currentBrokerController];
+        [brokerController completeAcquireTokenWithResult:nil error:error];
+    }
+}
+
+#endif
+
++ (void)stopTrackingAppState
+{
+    [[NSNotificationCenter defaultCenter] removeObserver:self
+                                                    name:UIApplicationDidBecomeActiveNotification
+                                                  object:nil];
+
+    [[NSNotificationCenter defaultCenter] removeObserver:self
+                                                    name:UIApplicationWillEnterForegroundNotification
+                                                  object:nil];
+}
+
+#pragma mark - Complete request
+
+- (BOOL)completeAcquireTokenWithResult:(MSIDTokenResult *)tokenResult error:(NSError *)error
+{
+    // TODO: stop telemetry event
+
+    [self.class setCurrentBrokerController:nil];
+    [self.class stopTrackingAppState];
+
+    if (self.requestCompletionBlock)
+    {
+        MSIDRequestCompletionBlock requestCompletion = [self.requestCompletionBlock copy];
+        self.requestCompletionBlock = nil;
```

Comment:
Do we want to put the above two lines into one synchronized block to avoid race condition?

Reply:
Let me check

Reply:
Done